### ISSUE TITLE
Idea: include generic-persistence in your review

### ISSUE THREAD

**Issue by thma** [Zelenya/elephants#3]
I'd love to see [generic-persistence](https://github.com/thma/generic-persistence) included in this review.

full disclosure: I'm the author of that framework.

**Comment by Zelenya** [comment]
I need a little break from postgres 😅 
I'd be more than happy to take a Pull Request or a draft in the meantime (I hope there are enough instructions on starting everything, but I'm here to help).

**Comment by thma** [comment]
That's fair!
I'll have a look at that an will come back with a PR.

**Comment by thma** [comment]
Over the weekend I've set up an Elephants.GenericPersistence module which implements your sample scenarios.

The code looks quite similar to the PersistentEsqueleto solution. Major differences:
- You don't need template Haskell to define Entities. 
- Entities are just normal record data types that implement the Entity type class. All Sql code is generated by using GHC Generics.
- There is a simple query API that allows to define where clauses for select statements. But there is no general typesafe SQL DSL. So whenever needed (for example for your tricky join statement) you will simply use a normal SQL statement, perform a select and use GenericPersistence to create Entity Objects from the resulting rows.
- There are two APIs provided: One which simply operates in the IO Monad and may throw PersistenceExceptions. (that's the one I have used in Elephants.GenericPersistence) 
A second one provides total functions that return IO (Either PersistenceException ...). This version is handy when writing servant handlers for example.

Personally I think that GenericPersistence sits in an attractive spot in the design space, because it provides basic persistence operations for Entities in quite a terse and also typesafe way. On the other hand it does not try to re-invent SQL.

I did not want to interfere with your review in the readme.md, so I did not write anything there. But if you like I could do this as well. just let me know.
